src/client/datascience/notebook/notebookSerliazer.ts

```diff
@@ -0,0 +1,103 @@
+ * Its up to extension authors to read the files and return it in a format that VSCode understands.
+ * Same with the cells and cell output.
+ */
+@injectable()
+export class NotebookSerializer implements VSCNotebookSerializer {
+    constructor(
+        @inject(NotebookCellLanguageService) private readonly cellLanguageService: NotebookCellLanguageService
+    ) {}
+    public deserializeNotebook(content: Uint8Array, _token: CancellationToken): NotebookData {
+        const contents = Buffer.from(content).toString();
+        const json = contents ? (JSON.parse(contents) as Partial<nbformat.INotebookContent>) : {};
+        traceInfoIf(IS_CI_SERVER, `NotebookJSON ${JSON.stringify(json)}`);
+
+        // Then compute indent. It's computed from the contents
+        const indentAmount = contents ? detectIndent(contents).indent : ' ';
+
+        // Then save the contents. We'll stick our cells back into this format when we save
+        if (json) {
+            // Log language or kernel telemetry
+            sendLanguageTelemetry(json);
+        }
+        const preferredCellLanguage = this.cellLanguageService.getPreferredLanguage(json?.metadata);
+        traceInfoIf(IS_CI_SERVER, `Preferred language in deserializer ${preferredCellLanguage}`);
+        // Ensure we always have a blank cell.
+        if ((json?.cells || []).length === 0) {
+            json.cells = [
+                {
+                    cell_type: 'code',
+                    execution_count: null,
+                    metadata: {},
+                    outputs: [],
+                    source: ''
+                }
+            ];
+        }
+        // For notebooks without metadata default the language in metadata to the preferred language.
+        if (!json.metadata || (!json.metadata.kernelspec && !json.metadata.language_info)) {
+            json.metadata = json?.metadata || { orig_nbformat: defaultNotebookFormat.major };
+            json.metadata.language_info = json.metadata.language_info || { name: preferredCellLanguage };
+        }
+        const data = notebookModelToVSCNotebookData(
+            { ...json, cells: [] },
+            json?.cells || [],
+            preferredCellLanguage,
+            json || {}
+        );
+        data.metadata = data.metadata || {};
+        data.metadata.indentAmount = indentAmount;
+
+        return data;
+    }
+    public serializeNotebookDocument(data: NotebookDocument): string {
+        return this.serialize(data);
+    }
+    public serializeNotebook(data: NotebookData, _token: CancellationToken): Uint8Array {
+        return Buffer.from(this.serialize(data), 'utf-8');
+    }
+    private serialize(data: NotebookDocument | NotebookData): string {
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const notebookContent: Partial<nbformat.INotebookContent> = (data.metadata?.custom as any) || {};
```

Comment:
When we serialize, is the kernel information already in the NotebookDocument|Data?

Reply:
yes

Reply:
when we select kernels, we edit metadata